### PR TITLE
smarter test_package version, user, channel management

### PR DESCRIPTION
**Pull request by memsharded**
Please feedback.

### CODE REVIEW

File: conans/client/command.py [hunk 1]
```diff
@@ -427,7 +432,12 @@ def export(self, *args):
                                  'Use for testing purposes only')
         parser.add_argument("--file", "-f", help="specify conanfile filename")
         args = parser.parse_args(*args)
-        return self._conan.export(user=args.user, path=args.path, keep_source=args.keep_source, filename=args.file)
+        try:
+            user, channel = args.user.split("/")
```
Comment (lasote) [inline conans/client/command.py]: Why keep the -a and -c if you are going to admit composed string in -a?

Comment (lasote) [inline conans/client/command.py]: Actually -c seems to be unused, right?

File: conans/test/command/test_package_test.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    requires = "Hello/0.2@user/channel"
+    def test(self):
+        pass
+'''
+        client = TestClient()
+        client.save({CONANFILE: conanfile,
+                     "test_package/conanfile.py": test_conanfile})
+        client.run("test_package")
+        self.assertIn("Hello/0.1@user/channel: Configuring sources", client.user_io.out)
+        self.assertIn("Hello/0.1@user/channel: Generated conaninfo.txt", client.user_io.out)
```
Comment (lasote) [inline conans/test/command/test_package_test.py]: Uf, black magic... requiring 0.2 but magically retrieved 0.1

**Comment by memsharded** [comment]
As you can see, there is a test with several requires declared. Now it works even if the required to the tested package is not the first one. In the previous implementation, it was hardcoded that the requirement should be the first one.

File: conans/test/command/test_package_test.py (hunk 2 above)
Comment (memsharded) [inline conans/test/command/test_package_test.py]: Yes, but the opposite is even more confusing. If you run this with 0.2 and have 0.1 in the root conanfile, the behavior is:
- Try to get 0.2 in the local cache, if it is there, use it!
- If not, try to get 0.2 from remotes, in order, when found, download and use it.
At the end, you are testing another package recipe different than the one you have!

File: conans/client/conan_api.py [hunk 3]
```diff
@@ -202,32 +201,41 @@ def test_package(self, path=None, profile_name=None, settings=None, options=None
...
-            first_dep = reqs[0][1].conan_reference
-        except Exception:
-            raise ConanException("Unable to retrieve first requirement of test conanfile.py")
+            if hasattr(test_conanfile, "requirements"):
+                test_conanfile.requirements()
+        except Exception as e:
+            raise ConanException("Error in test_package/conanfile.py requirements(). %s" % str(e))
+
+        requirement = test_conanfile.requires.get(package_name)
+        if requirement:
```
Comment (lasote) [inline conans/client/conan_api.py]: A warning if the requirement is specified to teach the new way?

File: conans/client/manager.py [hunk 4]
```diff
@@ -169,6 +166,16 @@ def _get_conanfile_object(self, loader, reference_or_path, conanfile_filename, c
 
         return conanfile
 
+    def _inject_require(self, conanfile, inject_require):
+        """ test_package functionality requires injecting the tested package as requirement
+        before running the install
+        """
+        require = conanfile.requires.get(inject_require.name)
+        if require:
+            require.conan_reference = require.range_reference = inject_require
```
Comment (lasote) [inline conans/client/manager.py]: Why is needed to assign the reference if the reference is already there?

Comment (memsharded) [inline conans/client/manager.py]: It might be pointing to a different user/channel

File: conans/client/conan_api.py (hunk 3 above)
Comment (memsharded) [inline conans/client/conan_api.py]: It depends on the approach we are following, if it is going to be the standard, then yes.

File: conans/client/manager.py (hunk 4 above)
Comment (lasote) [inline conans/client/manager.py]: ok

**Pull request merged**